### ISSUE TITLE
Phase 48.5: ContinualLearningOrchestrator — Unified Lifelong Learning Pipeline & Stability-Plasticity Balance

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#942]
### Overview & Motivation

The ContinualLearningOrchestrator is the capstone component of Phase 48, unifying all continual learning strategies — regularization-based (48.1), architecture-based (48.2), replay-based (48.3), and task detection (48.4) — into a coherent lifelong learning pipeline. It manages the **stability-plasticity dilemma**: maintaining knowledge of past tasks (stability) while retaining the ability to learn new ones (plasticity).

The orchestrator dynamically selects and combines strategies based on the detected learning scenario, available resources, and the nature of incoming tasks.

### Architecture Components

```
ContinualLearningOrchestrator
├── StrategySelector
│   ├── ScenarioClassifier           # Identify Task-IL, Domain-IL, or Class-IL
│   ├── StrategyRecommender          # Recommend optimal strategy for scenario
│   ├── HybridStrategyComposer      # Combine multiple strategies (EWC + Replay)
│   ├── ResourceAwareSelector        # Account for memory/compute constraints
│   └── AdaptiveStrategySwitcher     # Switch strategy mid-stream if needed
├── StabilityPlasticityManager
│   ├── BalanceMonitor               # Track stability-plasticity tradeoff
│   ├── PlasticityDecayScheduler     # Reduce plasticity as tasks accumulate
│   ├── StabilityEnforcer            # Ensure minimum retention thresholds
│   ├── AdaptiveRegularization       # Auto-tune lambda based on forgetting rate
│   └── CriticalPeriodDetector       # Identify optimal learning windows
├── TaskLifecycleManager
│   ├── TaskInitializer              # Initialize CL state for new task
│   ├── TaskTrainer                  # Training loop with CL augmentation
│   ├── TaskConsolidator             # Post-task consolidation (Fisher, exemplars, etc.)
│   ├── TaskEvaluator                # Evaluate all tasks after each learning phase
│   └── TaskRetirer                  # Archive old task state when resources are low
├── EvaluationEngine
│   ├── BackwardTransferMetric       # BWT = (1/T-1) Σ (R_{T,j} - R_{j,j})
│   ├── ForwardTransferMetric        # FWT = (1/T-1) Σ (R_{j-1,j} - b_j)
│   ├── ForgettingMeasure            # Max forgetting per task
│   ├── AverageAccuracyTracker       # Mean accuracy across all seen tasks
│   ├── LearningCurveAnalyzer        # Track per-task learning dynamics
│   └── AccuracyMatrix               # R[i][j] = accuracy on task j after learning task i
├── PipelineCoordinator
│   ├── DataStreamManager            # Manage continuous data streams
│   ├── BatchComposer                # Compose training batches with replay
│   ├── LossAggregator               # Combine task loss + regularization + replay
│   ├── GradientProcessor            # Apply GEM/A-GEM constraints + EWC penalty
│   └── CheckpointOrchestrator       # Coordinate model checkpointing across tasks
└── MonitoringDashboard
    ├── TaskTimelineVisualizer        # Timeline of tasks and boundaries
    ├── ForgettingHeatmap             # Per-task forgetting over time
    ├── StrategyEffectivenessTracker  # Compare strategy performance
    ├── ResourceUsageMonitor          # Memory, compute, and storage tracking
    └── ContinualLearningReport      # Comprehensive CL performance report
```

### Key Algorithms & Data Structures

**Unified Training Loop:**
```python
class ContinualLearningOrchestrator:
    def train_task(self, task_id, dataloader):
        self.task_lifecycle.initialize(task_id)
        strategy = self.strategy_selector.select(
            scenario=self.scenario_classifier.classify(dataloader),
            resources=self.resource_monitor.available(),
            task_similarity=self.task_detector.estimate_similarity(task_id)
        )

        for epoch in range(self.config.epochs_per_task):
            for batch in dataloader:
                # Mix with replay if applicable
                mixed_batch = self.replay_buffer.mix_batch(batch)
                # Forward pass
                logits = self.model(mixed_batch.x)
                # Task-specific loss
                task_loss = self.criterion(logits, mixed_batch.y)
                # Regularization penalty (EWC/SI/MAS)
                reg_penalty = self.consolidator.compute_penalty(self.model)
                # Total loss
                total_loss = task_loss + self.lambda_reg * reg_penalty
                total_loss.backward()
                # GEM gradient projection if applicable
                if strategy.uses_gem:
                    self.gem_projector.project_gradients(self.model)
                self.optimizer.step()

        # Post-task consolidation
        self.consolidator.consolidate_task(task_id)
        self.replay_buffer.update(task_id, dataloader)
        self.expander.freeze_column(task_id)
        # Evaluate on all tasks
        return self.evaluate_all_tasks()
```

**Accuracy Matrix & Continual Learning Metrics:**
```python
class EvaluationEngine:
    def compute_metrics(self, R: np.ndarray):
        """R[i][j] = accuracy on task j after training on task i"""
        T = R.shape[0]
        # Average accuracy after last task
        avg_acc = R[-1].mean()
        # Backward transfer
        bwt = sum(R[-1, j] - R[j, j] for j in range(T-1)) / (T - 1)
        # Forward transfer
        fwt = sum(R[i-1, i] - self.baselines[i] for i in range(1, T)) / (T - 1)
        # Average forgetting
        forgetting = sum(
            max(R[k, j] for k in range(j, T)) - R[-1, j]
            for j in range(T-1)
        ) / (T - 1)
        return CLMetrics(avg_acc=avg_acc, bwt=bwt, fwt=fwt, forgetting=forgetting)
```

**Adaptive Strategy Selection:**
```python
class StrategySelector:
    def select(self, scenario, resources, task_similarity):
        if scenario == "task_il":
            # Task identity known — architecture-based methods work best
            if resources.memory > self.thresholds.high:
                return HybridStrategy(progressive_nets=True, ewc=True)
            return ProgressiveNetsStrategy()
        elif scenario == "class_il":
            # Hardest scenario — replay is essential
            if resources.memory > self.thresholds.high:
                return HybridStrategy(replay=True, ewc=True, dark_replay=True)
            return ReplayStrategy(method="gem", budget=resources.memory // 2)
        else:  # domain_il
            if task_similarity > 0.8:
                return EWCStrategy(lambda_=1000)  # Similar domains, light regularization
            return HybridStrategy(ewc=True, replay=True)
```

### API Surface

```python
class ContinualLearningOrchestrator:
    def __init__(self, config: CLConfig): ...
    def register_model(self, model: nn.Module) -> None: ...
    def train_task(self, task_id: str, dataloader: DataLoader) -> TaskResult: ...
    def train_stream(self, data_stream: DataStream) -> StreamResult: ...
    def evaluate_all_tasks(self) -> AccuracyMatrix: ...
    def get_cl_metrics(self) -> CLMetrics: ...
    def select_strategy(self, scenario: str = "auto") -> Strategy: ...
    def set_strategy(self, strategy: Strategy) -> None: ...
    def get_forgetting_report(self) -> ForgettingReport: ...
    def get_stability_plasticity_balance(self) -> SPBalance: ...
    def checkpoint(self) -> CLCheckpoint: ...
    def restore(self, checkpoint: CLCheckpoint) -> None: ...
    def get_task_timeline(self) -> TaskTimeline: ...
    def export_report(self, format: str = "json") -> str: ...
```

### Testing Strategy

- End-to-end pipeline: train 5+ sequential tasks, verify positive BWT and near-zero forgetting
- Strategy selection: verify correct strategy chosen for each scenario type
- Hybrid strategy: verify combined EWC+Replay outperforms either alone
- Accuracy matrix correctness: R[i][j] matches independent evaluation
- Adaptive lambda: verify regularization strength increases with accumulated tasks
- Resource constraints: verify graceful degradation under memory pressure
- Checkpoint/restore: verify identical metrics after restoration
- Benchmark suite: Split-MNIST, Permuted-MNIST, Split-CIFAR-10, Split-CIFAR-100

### References

1. De Lange et al. (2021) — *A continual learning survey: Defying forgetting in classification tasks*
2. van de Ven & Tolias (2019) — *Three scenarios for continual learning*
3. Kirkpatrick et al. (2017) — *Overcoming catastrophic forgetting in neural networks*
4. Lopez-Paz & Ranzato (2017) — *Gradient Episodic Memory for Continual Learning*
5. Rusu et al. (2016) — *Progressive Neural Networks*
6. Buzzega et al. (2020) — *Dark Experience for General Continual Learning*
7. Hsu et al. (2018) — *Re-evaluating Continual Learning Scenarios*
8. Farquhar & Gal (2019) — *A Unifying Bayesian View of Continual Learning*
9. Parisi et al. (2019) — *Continual Lifelong Learning with Neural Networks: A Review*
10. Lesort et al. (2020) — *Continual Learning for Robotics: Definition, Framework, Learning Strategies*

Part of **Phase 48: Continual Learning & Catastrophic Forgetting Prevention**
Discussion: #937

**Comment by web3guru888** [comment]
## 🚀 Phase 49 Launched: Program Synthesis & Automated Code Generation

Phase 48 (Continual Learning & Catastrophic Forgetting Prevention) foundations enable Phase 49's program synthesis pipeline:

**Sub-phases:**
- **49.1** SpecificationParser — Formal specification parsing & NL-to-spec translation (#954)
- **49.2** SynthesisEngine — Enumerative, deductive & neural-guided program synthesis (#955)
- **49.3** CodeTransformer — Transformer-based code generation & program induction (#956)
- **49.4** VerificationEngine — Formal verification & counterexample-guided refinement (#957)
- **49.5** ProgramSynthesisOrchestrator — End-to-end synthesis pipeline & multi-strategy coordination (#958)

Planning: #953 | Wiki: 297 pages | Discussions: 660+
